File: js/src/memory/recorder.ts
/**
 * Background capture loop for screen memory — mirrors Python's
 * opendesk/memory/recorder.py.
 *
 * Every `interval_seconds` the recorder:
 *   1. checks the shared pause flag (hotkey / tool / CLI),
 *   2. reads the frontmost app + window title and consults the deny list,
 *   3. grabs the screen,
 *   4. skips the frame when it is visually identical to the previous one,
 *   5. runs OCR locally, builds a JPEG thumbnail, and stores both,
 *   6. periodically enforces retention and the storage cap.
 *
 * Nothing leaves the machine.
 */

import {
  MemoryConfig,
  clearDaemonState,
  clearPause,
  getPause,
  loadConfig,
  setPause,
  writeDaemonState,
} from "./config.js";
import { frontmost as detectFrontmost, type Frontmost } from "./frontmost.js";
import { frameSignature, makeThumbnail, signatureDelta } from "./image.js";
import { availableBackend, createOcrEngine, ensureVisionHelper, type OcrEngine, type OcrPreference } from "./ocr.js";
import { MemoryStore } from "./store.js";

export type CaptureFn = () => Promise<Buffer>;
export type OcrFn = (png: Buffer) => Promise<string>;
export type FrontmostFn = () => Promise<Frontmost>;

export interface RecorderOptions {
  home?: string;
  config?: MemoryConfig;
  store?: MemoryStore;
  /** Screen capture; defaults to screenshot-desktop. */
  capture?: CaptureFn;
  /** OCR; defaults to the best local engine (see ocr.ts). */
  ocr?: OcrFn;
  /** Frontmost app/title probe; defaults to the platform probe. */
  frontmost?: FrontmostFn;
  /** CLI override that survives config reloads. */
  intervalOverride?: number;
  log?: (msg: string) => void;
}

async function defaultCapture(): Promise<Buffer> {
  const mod = await import("screenshot-desktop");
  return mod.default({ format: "png" });
}

export class ScreenMemoryRecorder {
  readonly home?: string;
  config: MemoryConfig;
  readonly store: MemoryStore;
  private capture: CaptureFn;
  private ocrFn?: OcrFn;
  private engine: OcrEngine | null = null;
  private engineTried = false;
  private probe: FrontmostFn;
  private intervalOverride?: number;
  private log: (msg: string) => void;
  private lastSig: number[] | null = null;
  private ticks = 0;
  private _captured = 0;
  private _skipped = 0;
  private _lastStatus = "idle";
  private ocrWarned = false;
  private stopping = false;
  private wake: (() => void) | null = null;
  private hotkey: { stop(): void } | null = null;

  constructor(opts: RecorderOptions = {}) {
    this.home = opts.home;
    this.config = opts.config ?? loadConfig(opts.home);
    this.intervalOverride = opts.intervalOverride;
    if (this.intervalOverride) this.config.interval_seconds = this.intervalOverride;
    this.store = opts.store ?? new MemoryStore(opts.home);
    this.capture = opts.capture ?? defaultCapture;
    this.ocrFn = opts.ocr;
    this.probe = opts.frontmost ?? detectFrontmost;
    this.log = opts.log ?? ((m) => process.stderr.write(`[memory] ${m}\n`));
  }

  get captured(): number { return this._captured; }
  get skipped(): number { return this._skipped; }
  get lastStatus(): string { return this._lastStatus; }

  /** Pick up deny-list / interval edits made by other processes. */
  reloadConfig(): void {
    const fresh = loadConfig(this.home);
    if (this.intervalOverride) fresh.interval_seconds = this.intervalOverride;
    this.config = fresh;
  }

  // -- one tick ------------------------------------------------------------

  async tick(): Promise<string> {
    this.ticks++;
    if (this.ticks % 10 === 0) this.reloadConfig();

    if (getPause(this.home) !== null) return this.done("paused");

    const { app, title } = await this.frontmost();
    if (this.config.isDenied(app, title)) return this.done(`denied:${app || title}`);

    let png: Buffer;
    try {
      png = await this.capture();
    } catch (e) {
      this.log(`capture failed: ${e}`);
      return this.done(`error:capture:${e}`);
    }

    let sig: number[] | null = null;
    try {
      sig = await frameSignature(png);
    } catch (e) {
      this.log(`signature failed: ${e}`);
    }
    if (sig && this.lastSig && signatureDelta(sig, this.lastSig) < this.config.dedupe_threshold) {
      return this.done("duplicate");
    }
    this.lastSig = sig;

    let text = "";
    try {
      text = await this.recognize(png);
    } catch (e) {
      if (!this.ocrWarned) {
        this.log(`OCR failed — frames will be stored without text: ${e}`);
        this.ocrWarned = true;
      }
      text = "";
    }

    let thumb: Buffer | null = null;
    let tw = 0;
    let th = 0;
    try {
      const t = await makeThumbnail(png, this.config.thumbnail_width, this.config.thumbnail_quality);
      thumb = t.jpeg; tw = t.width; th = t.height;
    } catch (e) {
      this.log(`thumbnail failed: ${e}`);
    }

    this.store.add({ ts: Date.now() / 1000, app, title, text, thumb, width: tw, height: th });
    this._captured++;
    if (this._captured % 20 === 1) this.housekeep();
    return this.done("captured");
  }

  /** Apply retention + storage cap. */
  housekeep(): { expired: number; capped: number } {
    const expired = this.store.enforceRetention(this.config.retention_days);
    const capped = this.store.enforceCap(this.config.storage_cap_mb * 1024 * 1024);
    if (expired || capped) this.log(`housekeeping: removed ${expired} expired, ${capped} over-cap frames`);
    return { expired, capped };
  }

  private async recognize(png: Buffer): Promise<string> {
    if (this.ocrFn) return this.ocrFn(png);
    if (!this.engine && !this.engineTried) {
      this.engineTried = true;
      this.engine = await createOcrEngine(this.config.ocr_backend as OcrPreference, this.home);
      if (!this.engine) throw new Error("no OCR backend available");
    }
    if (!this.engine) throw new Error("no OCR backend available");
    return this.engine.recognize(png);
  }

  private async frontmost(): Promise<Frontmost> {
    try {
      const f = await this.probe();
      const app = (f.app ?? "").trim();
      const title = (f.title ?? "").trim();
      return { app, title: title === app ? "" : title };
    } catch {
      return { app: "", title: "" };
    }
  }

  private done(status: string): string {
    this._lastStatus = status;
    if (status !== "captured") this._skipped++;
    writeDaemonState(this.home, {
      status, captured: this._captured, skipped: this._skipped, interval: this.config.interval_seconds,
    });
    return status;
  }

  // -- loop ----------------------------------------------------------------

  async run(): Promise<void> {
    this.stopping = false;
    writeDaemonState(this.home, { status: "starting", captured: 0, skipped: 0, interval: this.config.interval_seconds });
    await this.startHotkey();
    try {
      while (!this.stopping) {
        const started = Date.now();
        try {
          await this.tick();
        } catch (e) {
          this.log(`tick failed: ${e}`);
          this.done(`error:${e}`);
        }
        const elapsed = (Date.now() - started) / 1000;
        const delay = Math.max(1, this.config.interval_seconds - elapsed);
        if (this.stopping) break;
        await new Promise<void>((resolve) => {
          const t = setTimeout(() => { this.wake = null; resolve(); }, delay * 1000);
          this.wake = () => { clearTimeout(t); this.wake = null; resolve(); };
        });
      }
    } finally {
      this.stopHotkey();
      if (this.engine) await this.engine.close();
      clearDaemonState(this.home);
      this.store.close();
    }
  }

  stop(): void {
    this.stopping = true;
    this.wake?.();
  }

  // -- pause hotkey ----------------------------------------------------------

  /** Flip the pause flag.  Returns the new paused state. */
  togglePause(): boolean {
    if (getPause(this.home) !== null) {
      clearPause(this.home);
      console.log(`[${now()}] screen memory resumed`);
      return false;
    }
    setPause(this.home, { reason: "hotkey" });
    console.log(`[${now()}] screen memory paused — press the hotkey again to resume`);
    return true;
  }

  private async startHotkey(): Promise<void> {
    const combo = (this.config.pause_hotkey ?? "").trim();
    if (!combo) return;
    try {
      const { registerHotkey } = await import("./hotkey.js");
      this.hotkey = await registerHotkey(combo, () => this.togglePause());
      this.log(`pause hotkey armed: ${combo}`);
    } catch (e) {
      this.log(`pause hotkey unavailable (${(e as Error).message}). Use 'opendesk-js memory pause' instead.`);
      this.hotkey = null;
    }
  }

  private stopHotkey(): void {
    try { this.hotkey?.stop(); } catch { /* ignore */ }
    this.hotkey = null;
  }
}

// ---------------------------------------------------------------------------
// Daemon entry point
// ---------------------------------------------------------------------------

export async function startDaemon(opts: { home?: string; interval?: number } = {}): Promise<void> {
  const config = loadConfig(opts.home);
  const backend = availableBackend(config.ocr_backend as OcrPreference, opts.home);
  if (backend === "macos-vision") {
    console.log("Preparing the macOS Vision OCR helper (one-time compile, may take a minute)…");
    await ensureVisionHelper(opts.home);
  }
  if (!backend) {
    process.stderr.write(
      "WARNING: no OCR backend available — frames will be stored without text.\n" +
      "  tesseract.js is bundled; on macOS install Xcode command-line tools for the faster Vision engine.\n",
    );
  }

  const recorder = new ScreenMemoryRecorder({ home: opts.home, config, intervalOverride: opts.interval });
  const cfg = recorder.config;
  console.log("opendesk screen memory (js)");
  console.log(`  store:      ${recorder.store.dir}`);
  console.log(`  interval:   every ${cfg.interval_seconds}s`);
  console.log(`  cap:        ${cfg.storage_cap_mb} MB, retention ${cfg.retention_days} days`);
  console.log(`  deny list:  ${cfg.deny_apps.join(", ") || "(empty)"}`);
  console.log(`  OCR:        ${backend ?? "unavailable"}`);
  console.log(`  pause key:  ${cfg.pause_hotkey || "(disabled)"}`);
  if (getPause(opts.home) !== null) {
    console.log("  state:      PAUSED (resume with `opendesk-js memory resume` or the hotkey)");
  }
  console.log("Ctrl-C to stop.\n");

  const stop = () => recorder.stop();
  process.on("SIGINT", stop);
  process.on("SIGTERM", stop);
  await recorder.run();
  console.log(`\nStopped. ${recorder.captured} frames captured this run.`);
}

function now(): string {
  return new Date().toTimeString().slice(0, 8);
}

File: python/opendesk/memory/recorder.py
"""Background capture loop for screen memory.

Every ``interval_seconds`` the recorder:

1. Checks the shared pause flag (hotkey / ``memory(action=pause)`` / CLI).
2. Reads the frontmost app + window title and consults the deny list.
3. Grabs the screen through the :class:`~opendesk.computer.Computer`.
4. Skips the frame when it is visually identical to the previous one.
5. Runs OCR locally, builds a JPEG thumbnail, and writes both to the
   :class:`~opendesk.memory.store.MemoryStore`.
6. Periodically enforces retention and the storage cap.

Nothing leaves the machine: capture, OCR, and storage are all local.
"""

from __future__ import annotations

import asyncio
import io
import logging
import signal
import sys
import time
from pathlib import Path
from typing import Any, Callable, Optional

from opendesk.memory.config import (
    MemoryConfig,
    clear_daemon_state,
    clear_pause,
    get_pause,
    load_config,
    set_pause,
    write_daemon_state,
)
from opendesk.memory.store import MemoryStore

log = logging.getLogger("opendesk.memory")

OCRFunc = Callable[[bytes], str]


# ---------------------------------------------------------------------------
# Image helpers (pure functions, unit-testable)
# ---------------------------------------------------------------------------


def make_thumbnail(png_bytes: bytes, width: int = 480, quality: int = 60) -> tuple[bytes, int, int]:
    """Downscale a PNG to a JPEG thumbnail.  Returns ``(jpeg, w, h)``."""
    from PIL import Image  # type: ignore[import-not-found]

    img = Image.open(io.BytesIO(png_bytes)).convert("RGB")
    if img.width > width:
        ratio = width / img.width
        img = img.resize((width, max(1, int(img.height * ratio))), Image.LANCZOS)
    buf = io.BytesIO()
    img.save(buf, format="JPEG", quality=quality, optimize=True)
    return buf.getvalue(), img.width, img.height


def frame_signature(png_bytes: bytes, size: int = 16) -> list[int]:
    """Tiny grayscale fingerprint used for duplicate detection."""
    from PIL import Image  # type: ignore[import-not-found]

    img = Image.open(io.BytesIO(png_bytes)).convert("L").resize((size, size), Image.BILINEAR)
    return list(img.getdata())


def signature_delta(a: Optional[list[int]], b: Optional[list[int]]) -> float:
    """Mean absolute difference between two signatures, normalised to 0–1."""
    if not a or not b or len(a) != len(b):
        return 1.0
    total = sum(abs(x - y) for x, y in zip(a, b))
    return total / (255.0 * len(a))


def _default_ocr(png_bytes: bytes) -> str:
    from opendesk.computer.ocr import ocr_image
    return ocr_image(png_bytes)


def _ocr_failed(text: str) -> bool:
    return (
        text.startswith("OCR not available")
        or text.startswith("OCR error")
        or text.startswith("pytesseract error")
    )


# ---------------------------------------------------------------------------
# Recorder
# ---------------------------------------------------------------------------


class ScreenMemoryRecorder:
    """Drives one capture tick at a time.  :meth:`run` loops forever."""

    def __init__(
        self,
        home: Optional[Path] = None,
        *,
        computer: Any = None,
        config: Optional[MemoryConfig] = None,
        store: Optional[MemoryStore] = None,
        ocr: Optional[OCRFunc] = None,
        interval_override: Optional[float] = None,
    ) -> None:
        self.home = home
        self.config = config or load_config(home)
        self._interval_override = float(interval_override) if interval_override else None
        if self._interval_override:
            self.config.interval_seconds = self._interval_override
        self.store = store or MemoryStore(home)
        self._computer = computer
        self._ocr: OCRFunc = ocr or _default_ocr
        self._last_sig: Optional[list[int]] = None
        self._ticks = 0
        self._captured = 0
        self._skipped = 0
        self._last_status = "idle"
        self._ocr_warned = False
        self._stop = asyncio.Event()
        self._hotkey_listener: Any = None

    # -- accessors -----------------------------------------------------

    @property
    def computer(self) -> Any:
        if self._computer is None:
            from opendesk.computer.local import LocalComputer
            self._computer = LocalComputer()
        return self._computer

    @property
    def captured(self) -> int:
        return self._captured

    @property
    def skipped(self) -> int:
        return self._skipped

    @property
    def last_status(self) -> str:
        return self._last_status

    def reload_config(self) -> None:
        """Pick up deny-list / interval edits made by other processes."""
        fresh = load_config(self.home)
        # Keep a CLI interval override sticky across reloads.
        if self._interval_override:
            fresh.interval_seconds = self._interval_override
        self.config = fresh

    # -- one tick --------------------------------------------------------

    async def tick(self) -> str:
        """Perform one capture attempt.  Returns a short status string."""
        self._ticks += 1
        if self._ticks % 10 == 0:
            self.reload_config()

        pause = get_pause(self.home)
        if pause is not None:
            return self._done("paused")

        app, title = await self._frontmost()
        if self.config.is_denied(app, title):
            return self._done(f"denied:{app or title}")

        try:
            pixmap = await self.computer.capture()
        except Exception as exc:
            log.warning("capture failed: %s", exc)
            return self._done(f"error:capture:{exc}")

        png = pixmap.data
        loop = asyncio.get_running_loop()
        try:
            sig = await loop.run_in_executor(None, frame_signature, png)
        except Exception as exc:
            log.warning("signature failed: %s", exc)
            sig = None
        if sig is not None and self._last_sig is not None:
            if signature_delta(sig, self._last_sig) < self.config.dedupe_threshold:
                return self._done("duplicate")
        self._last_sig = sig

        try:
            text = await loop.run_in_executor(None, self._ocr, png)
        except Exception as exc:
            text = f"OCR error: {exc}"
        if _ocr_failed(text):
            if not self._ocr_warned:
                log.warning("OCR unavailable — frames will be stored without text: %s", text.splitlines()[0])
                self._ocr_warned = True
            text = ""
        elif text == "(no text detected)":
            text = ""

        try:
            thumb, tw, th = await loop.run_in_executor(
                None, make_thumbnail, png,
                self.config.thumbnail_width, self.config.thumbnail_quality,
            )
        except Exception as exc:
            log.warning("thumbnail failed: %s", exc)
            thumb, tw, th = None, 0, 0

        frame = self.store.add(
            ts=time.time(), app=app, title=title, text=text,
            thumb=thumb, width=tw, height=th,
        )
        self._captured += 1
        log.debug("captured frame %s app=%r title=%r chars=%d", frame.id, app, title, len(text))

        if self._captured % 20 == 1:
            await loop.run_in_executor(None, self.housekeep)
        return self._done("captured")

    def housekeep(self) -> dict[str, int]:
        """Apply retention + storage cap.  Safe to call from any thread."""
        expired = self.store.enforce_retention(self.config.retention_days)
        capped = self.store.enforce_cap(self.config.storage_cap_mb * 1024 * 1024)
        if expired or capped:
            log.info("housekeeping: removed %d expired, %d over-cap frames", expired, capped)
        return {"expired": expired, "capped": capped}

    def _done(self, status: str) -> str:
        self._last_status = status
        if status != "captured":
            self._skipped += 1
        write_daemon_state(
            self.home,
            status=status,
            captured=self._captured,
            skipped=self._skipped,
            interval=self.config.interval_seconds,
        )
        return status

    async def _frontmost(self) -> tuple[str, str]:
        try:
            win = await self.computer.focused_window()
        except Exception:
            return "", ""
        if win is None:
            return "", ""
        app = (getattr(win, "app_name", "") or "").strip()
        title = (getattr(win, "title", "") or "").strip()
        if title == app:
            title = ""
        return app, title

    # -- loop --------------------------------------------------------------

    async def run(self) -> None:
        """Capture until :meth:`stop` is called."""
        write_daemon_state(self.home, status="starting", captured=0, skipped=0,
                           interval=self.config.interval_seconds)
        self._start_hotkey()
        try:
            while not self._stop.is_set():
                started = time.monotonic()
                try:
                    await self.tick()
                except Exception as exc:  # never let one bad tick kill the loop
                    log.exception("tick failed: %s", exc)
                    self._done(f"error:{exc}")
                elapsed = time.monotonic() - started
                delay = max(1.0, self.config.interval_seconds - elapsed)
                try:
                    await asyncio.wait_for(self._stop.wait(), timeout=delay)
                except asyncio.TimeoutError:
                    pass
        finally:
            self._stop_hotkey()
            clear_daemon_state(self.home)
            self.store.close()

    def stop(self) -> None:
        self._stop.set()

    # -- pause hotkey --------------------------------------------------------

    def toggle_pause(self) -> bool:
        """Flip the pause flag.  Returns the new paused state."""
        if get_pause(self.home) is not None:
            clear_pause(self.home)
            log.info("screen memory resumed (hotkey)")
            print(f"[{_now()}] screen memory resumed", flush=True)
            return False
        set_pause(self.home, reason="hotkey")
        log.info("screen memory paused (hotkey)")
        print(f"[{_now()}] screen memory paused — press the hotkey again to resume", flush=True)
        return True

    def _start_hotkey(self) -> None:
        combo = (self.config.pause_hotkey or "").strip()
        if not combo:
            return
        try:
            from pynput import keyboard  # type: ignore[import-not-found]
        except ImportError:
            log.warning(
                "pynput not installed — pause hotkey disabled. "
                "pip install 'opendesk[memory]' to enable it."
            )
            return
        try:
            self._hotkey_listener = keyboard.GlobalHotKeys({combo: self.toggle_pause})
            self._hotkey_listener.daemon = True
            self._hotkey_listener.start()
            log.info("pause hotkey armed: %s", combo)
        except Exception as exc:
            log.warning("could not register pause hotkey %r: %s", combo, exc)
            self._hotkey_listener = None

    def _stop_hotkey(self) -> None:
        if self._hotkey_listener is not None:
            try:
                self._hotkey_listener.stop()
            except Exception:
                pass
            self._hotkey_listener = None


# ---------------------------------------------------------------------------
# Daemon entry point
# ---------------------------------------------------------------------------


def start_daemon(home: Optional[Path] = None, *, interval: Optional[float] = None) -> None:
    """Blocking entry used by ``opendesk memory start``."""
    from opendesk.computer.ocr import available_backend, warm_up

    config = load_config(home)
    backend = available_backend()
    if backend == "macos-vision":
        print("Preparing the macOS Vision OCR helper (one-time compile, may take a minute)…", flush=True)
        warm_up()
    if backend is None:
        print(
            "WARNING: no OCR backend available — frames will be stored without text.\n"
            "  pip install pytesseract  (and install the tesseract binary)",
            file=sys.stderr,
        )

    recorder = ScreenMemoryRecorder(home, config=config, interval_override=interval)
    config = recorder.config
    store_dir = recorder.store.dir
    print("opendesk screen memory")
    print(f"  store:      {store_dir}")
    print(f"  interval:   every {config.interval_seconds:g}s")
    print(f"  cap:        {config.storage_cap_mb} MB, retention {config.retention_days} days")
    print(f"  deny list:  {', '.join(config.deny_apps) or '(empty)'}")
    print(f"  OCR:        {backend or 'unavailable'}")
    print(f"  pause key:  {config.pause_hotkey or '(disabled)'}")
    if get_pause(home) is not None:
        print("  state:      PAUSED (resume with `opendesk memory resume` or the hotkey)")
    print("Ctrl-C to stop.\n")

    async def _main() -> None:
        loop = asyncio.get_running_loop()

        def _request_stop(*_: Any) -> None:
            recorder.stop()

        for sig in (signal.SIGINT, signal.SIGTERM):
            try:
                loop.add_signal_handler(sig, _request_stop)
            except (NotImplementedError, RuntimeError):
                signal.signal(sig, _request_stop)
        await recorder.run()

    try:
        asyncio.run(_main())
    except KeyboardInterrupt:
        pass
    print(f"\nStopped. {recorder.captured} frames captured this run.")


def _now() -> str:
    import datetime
    return datetime.datetime.now().strftime("%H:%M:%S")
